File: jobs/2026-05-27__00-37-24/money_transfer_saga_ts__TukrZBJ/artifacts/user/myproject/src/activities.ts
import * as fs from 'fs/promises';
import { ApplicationFailure } from '@temporalio/activity';

const ACCOUNTS_FILE = '/workspace/accounts.json';

async function readAccounts(): Promise<Record<string, number>> {
  const data = await fs.readFile(ACCOUNTS_FILE, 'utf8');
  return JSON.parse(data);
}

async function writeAccounts(accounts: Record<string, number>): Promise<void> {
  await fs.writeFile(ACCOUNTS_FILE, JSON.stringify(accounts, null, 2));
}

export async function withdraw(account: string, amount: number): Promise<void> {
  console.log(`Withdrawing ${amount} from ${account}`);
  const accounts = await readAccounts();
  if (!accounts[account] && accounts[account] !== 0) {
    accounts[account] = 0;
  }
  accounts[account] -= amount;
  await writeAccounts(accounts);
}

export async function deposit(account: string, amount: number): Promise<void> {
  console.log(`Depositing ${amount} to ${account}`);
  if (account === 'B_FAIL') {
    throw ApplicationFailure.create({
      message: 'Intentional deposit failure for B_FAIL',
      nonRetryable: true,
    });
  }
  const accounts = await readAccounts();
  if (!accounts[account] && accounts[account] !== 0) {
    accounts[account] = 0;
  }
  accounts[account] += amount;
  await writeAccounts(accounts);
}

export async function refund(account: string, amount: number): Promise<void> {
  console.log(`Refunding ${amount} to ${account}`);
  const accounts = await readAccounts();
  if (!accounts[account] && accounts[account] !== 0) {
    accounts[account] = 0;
  }
  accounts[account] += amount;
  await writeAccounts(accounts);
}

File: jobs/2026-05-27__00-37-24/money_transfer_saga_py__JcvKYMs/artifacts/user/myproject/activities.py
import json
import os
from temporalio import activity
from temporalio.exceptions import ApplicationError

ACCOUNTS_FILE = "/workspace/accounts.json"

def _read_accounts():
    with open(ACCOUNTS_FILE, "r") as f:
        return json.load(f)

def _write_accounts(accounts):
    with open(ACCOUNTS_FILE, "w") as f:
        json.dump(accounts, f, indent=2)

@activity.defn
async def withdraw(from_account: str, amount: int):
    accounts = _read_accounts()
    if accounts.get(from_account, 0) < amount:
        raise ApplicationError(f"Insufficient funds in {from_account}")
    accounts[from_account] -= amount
    _write_accounts(accounts)
    return f"Withdrew {amount} from {from_account}"

@activity.defn
async def deposit(to_account: str, amount: int):
    if to_account == "B_FAIL":
        raise ApplicationError("Simulated deposit failure", non_retryable=True)
    
    accounts = _read_accounts()
    accounts[to_account] = accounts.get(to_account, 0) + amount
    _write_accounts(accounts)
    return f"Deposited {amount} to {to_account}"

@activity.defn
async def refund(from_account: str, amount: int):
    accounts = _read_accounts()
    accounts[from_account] = accounts.get(from_account, 0) + amount
    _write_accounts(accounts)
    return f"Refunded {amount} to {from_account}"
